src/guided_fuzzing_daemon/afl.py

```diff
@@ -0,0 +1,368 @@
+        corpus_seed.mkdir()
+        copy(seed, corpus_seed)
+
+        while opts.max_runtime > time() - start:
+            # check and restart subprocesses
+            for idx, proc in enumerate(procs):
+                if proc and proc.poll() is not None:
+                    print(f"afl-fuzz returned early: {proc.wait()}", file=sys.stderr)
+                    procs[idx] = proc = None
+                    stats.fields["instances"] -= 1  # type: ignore
+
+                if (
+                    proc is None
+                    # rate limit AFL++ instance launch to 1/minute
+                    and last_afl_start < time() - 60
+                    # don't launch secondary instances until main instance has finished
+                    # initializing
+                    and (not idx or (opts.corpus_out / "0" / "fuzzer_stats").exists())
+                ):
+                    if idx:
+                        cmd = ["-S", str(idx), "-p", choice(POWER_SCHEDS)]
+                    else:
+                        cmd = ["-M", "0"]
+                        if opts.afl_async_corpus:
+                            cmd.extend(("-F", str(opts.corpus_in)))
+                        if opts.afl_add_corpus:
+                            for additional_corpus in opts.afl_add_corpus:
+                                cmd.extend(("-F", str(additional_corpus)))
+
+                    # environment settings that apply to this instance
+                    this_env = envs[idx].copy()
+                    if opts.afl_async_corpus and not idx:
+                        this_env["AFL_IMPORT_FIRST"] = "1"
+                    if not idx:
+                        this_env["AFL_FINAL_SYNC"] = "1"
+
+                    # pylint: disable=consider-using-with
+                    procs[idx] = Popen(
+                        [
+                            str(afl_fuzz),
+                            "-t",
+                            str(opts.afl_timeout),
+                            *cmd,
+                            "-i",
+                            str(
+                                (
+                                    corpus_seed
+                                    if opts.afl_async_corpus
+                                    else opts.corpus_in
+                                ).resolve()
+                            ),
+                            "-o",
+                            str(opts.corpus_out.resolve()),
+                            "--",
+                            str(binary),
+                        ],
+                        text=True,
+                        env=this_env,
+                        stderr=STDOUT,
+                        stdout=log_tee.open_files[idx].handle,
+                    )
+                    last_afl_start = time()
+                    stats.fields["instances"] += 1  # type: ignore
+
+            # tee logs
+            log_tee.print()
+
+            # submit any crashes
+            if collector:
```

Comment:
Maybe add a comment noting the expected directory structure. This would be helpful in the chunk of code as a reference.